tests/apps/demultiplex/test_sample_models.py

```diff
@@ -0,0 +1,398 @@
+
+
+@pytest.mark.parametrize(
+    "lims_sample",
+    [
+        {"index": "GTCTACAC-GCCAAGGT", "sample_id": "ACC123"},
+        {"lane": 1, "sample_id": "ACC123"},
+        {"lane": 1, "index": "GTCTACAC-GCCAAGGT"},
+    ],
+    ids=["no sample id", "no index", "no lane"],
+)
+def test_validate_inputs_bcl_convert_sample_missing_attribute(lims_sample: dict):
+    """Test that validating a BCLConvert sample with a missing attribute fails."""
+    # GIVEN a raw LIMS sample without a mandatory attribute
+
+    # WHEN validating the sample
+    with pytest.raises(AttributeError) as exc_info:
+        FlowCellSampleBCLConvert.validate_inputs(lims_sample=lims_sample)
+
+    # THEN a pydantic validation error is raised
+    assert "validate_inputs" in str(exc_info.value)
+
+
+@pytest.mark.parametrize(
+    "lims_index, expected_index_1, expected_index2",
+    [
+        ("CCATTCGANNNNNNNNN-GTTGTCCG", "CCATTCGA", "GTTGTCCG"),
+        ("CGATAGCAGG-AATGCTACGA", "CGATAGCAGG", "AATGCTACGA"),
+        ("ACAGCAAG", "ACAGCAAG", ""),
+        ("ACAGCAAG-", "ACAGCAAG", ""),
+        ("ATAGCAGG-AATGCTAC\n", "ATAGCAGG", "AATGCTAC"),
+    ],
+    ids=[
+        "17-nt index1",
+        "equal-length indexes",
+        "single index",
+        "single index with dash",
+        "index with newline",
+    ],
+)
+def test_separate_indexes_dual_run(lims_index: str, expected_index_1: str, expected_index2: str):
+    """Test that parsing different kinds of dual-run raw indexes as index and index2 works."""
+    # GIVEN a sample sheet with a single sample on one lane
+    sample = FlowCellSampleBCLConvert(lane=1, index=lims_index, sample_id="ACC123")
+
+    # WHEN separating the index
+    sample.separate_indexes(is_run_single_index=False)
+
+    # THEN the index should be separated
+    assert sample.index == expected_index_1
+    assert sample.index2 == expected_index2
+
+
+def test_separate_indexes_single_run(bcl_convert_flow_cell_sample: FlowCellSampleBCLConvert):
+    """Test index2 is ignored when parsing a double index in a single index run."""
+    # GIVEN a sample with a double index
+
+    # WHEN separating the index
+    bcl_convert_flow_cell_sample.separate_indexes(is_run_single_index=True)
+
+    # THEN the index should be separated
+    assert bcl_convert_flow_cell_sample.index == "GTCTACAC"
+    assert bcl_convert_flow_cell_sample.index2 == ""
+
+
+@pytest.mark.parametrize(
+    "needs_reverse_complement, expected_index2",
+    [
+        (True, f"{INDEX_TWO_PAD_SEQUENCE}GCCAAGGT"),
+        (False, f"GCCAAGGT{INDEX_TWO_PAD_SEQUENCE}"),
+    ],
+    ids=["reverse complement", "no reverse complement"],
+)
+def test_pad_indexes_needs_padding(needs_reverse_complement: bool, expected_index2: str):
+    """Test that indexes that need to be padded are padded with and without reverse complement."""
+    # GIVEN an IndexSettings object
+    mock_index_settings = Mock(
+        spec=IndexSettings, should_i5_be_reverse_complimented=needs_reverse_complement
+    )
```

Comment:
Cool stuff, but why can we not just instantiate a real object?

Reply:
I guess this could be a real instance of IndexSettings. I reason that the focus of this test is to try different values of `should_i5_be_reverse_complimented`, without caring about the other attribute of the class, `are_i5_override_cycles_reverse_complemented`, and to create a real instance of `IndexSettings` would need a value for `are_i5_override_cycles_reverse_complemented`. That is why I chose to use a Mock. Do you think it should be a real instance? @Vince-janv

Reply:
I would probably just import the appropriate IndexSettings-constant and use that. It does not seem very cumbersome, and I feel mocking should be saved for more difficult scenarios

Reply:
good point, fixed